Add Lint checks for the new time API

lukstbit opened the pull request:
## Purpose / Description
This is an attempt to implement the Lint rules requested in #6951 . 

To implement the rules a good(and recent) source of information was [this excelent video](https://www.youtube.com/watch?v=jCmJWOkjbM0) (from the AndroidDevSummit) and the accompanying [repository](https://github.com/tikurahul/lint-experiments) for actual code examples. 
**Note**: To add new Lint rules to the project you'll have to use the available `com.ichi2.anki.lint.IssueRegistry` class. After creating the new rules, register them with the `IssueRegistry` by adding them to the returned list of issues in the `getIssues()` method!

##

Fixes #6951

## How Has This Been Tested?
All Lint rules have a passing test.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

Arthur-Milchior commented:
First, I want to thank you very much for this contribution. I find it incredible that a pull request by a new contributor closes a TODO that was added only yesterday because of a change I made without warning anybody in advance. I didn't know how to do lint test, I guess I should have done it at some time otherwise, you saved me a lot of time.

I've got two remarks, concerning the text. While technically, it's true that `SystemTime` is going to be used while running, in practice, that is not what we want the user to read. What we want is that they should use the `Collection`'s `getTime()`. In particular, they want to use the abstract class `Time`.

The risk is that otherwise they just do `new SystemTime().intTime()`, which would give a correct result and not exactly improve our goal.
Actually, I guess that an even better thing to do would be to forbid the creation of new system time, except in the place where we already use it. Speaking of which, I discovered we already have too much `new SystemTime` in the code base (I'll be deleting them in other PR)

Arthur-Milchior commented:
Second, I'd like to know why this choice of path. It seems strange to me to use the "src/main/java/com/ichi2/anki" path, as the rules applies to other subdirectory of ichi2

david-allison-1 reviewed (approved):
I don't think I have anything to add, this looks incredible, both from a speed and an implementation point of view. Thank you!

This isn't an area of Android/Java that I'm familiar with. It would be useful to link some documentation that you used in the PR description so I and others can better follow this.

Arthur-Milchior commented on lint-rules/src/main/java/com/ichi2/anki/lint/utils/LintUtils.java:
```diff
@@ -0,0 +1,32 @@
+        for (int i = 0; i < classes.size(); i++) {
+            final String className = classes.get(i).getName();
+            if (className.equals("Time") || className.equals("SystemTime")) {
+                isInAllowedClass = true;
```
Does it mean you area allowing the call of time methods in the class Time ? It does not seems right. The abstract class Time should not get any access to real time

lukstbit replied:
My assumption was that these classes should have direct access to Date, Calendar, System.currentTimeMillis() etc. Isn't this ok?

Arthur-Milchior replied:
The class `Time` need to be able to create calendar and date object. So I guess you're right in a way, even if they change the date later. However, they don't need to get access to real time through `currentTimeMillis`, since they get the time via `intTimeMS`, which is the only abstract overridden. `currentTimeMillis` is only called in `SystemTime`.

I guess ideally, this should allow the creation of calendar/date object only if they are linked to a specific time, and not to current time, but I don't know how to do it since it seems the calendar/date is first created with current time before getting changed.

Speaking of which, I'm surprised lint work, because the code still contains
```java
        Calendar sd = GregorianCalendar.getInstance();
        sd.setTimeInMillis(mCol.getSched().getDayCutoff() * 1000);
```
(which is perfectly valid since we don't actually uses current time here)

And 
```java
    public static Calendar reminderToCalendar(JSONObject reminder) {

        final Calendar calendar = Calendar.getInstance();

        calendar.set(Calendar.HOUR_OF_DAY, reminder.getJSONArray("time").getInt(0));
        calendar.set(Calendar.MINUTE, reminder.getJSONArray("time").getInt(1));
        calendar.set(Calendar.SECOND, 0);
        return calendar;
    }
```
Which is not valid since it uses the real day (I don't know how I missed it, my fault)

So I guess there is something I didn't understand in what your lint rules should detect.


I guess the best option is to have in Time a static method which, given a timestamp, generate the calendar; this way the Calendar creation will be in an authorized class. And the second example need to be changed to get `col.getTime`'s calendar, so the reminder is computed to the correct day (i don't know how I missed when I changed all times. I assume that since the hour was changed, I assumed it was okay to keep it in the code. How silly of mine)

Arthur-Milchior commented on lint-rules/src/main/java/com/ichi2/anki/lint/rules/DirectSystemTimeUsage.java:
```diff
@@ -0,0 +1,75 @@
+ * of using the time provided by the new SystemTime class.
+ */
+public class DirectSystemTimeUsage extends Detector implements SourceCodeScanner {
+
```
The name is slightly confusing here, since "SystemTime" is a class of AnkiDroid and this is not related to the class SystemTime

lukstbit replied:
Yes, that is confusing. Would `DirectSystemMillisUsage` be a better name?

mikehardy replied:
I recognize this is overly long, but it's also not something we'll have to type a lot, maybe going all the way and saying "DirectSystemCurrentTimeMillisUsage" - now it's completely explicit and no need to mentally translate

Arthur-Milchior reviewed (commented):
Once again, thanks a lot. Those are hopefully some small remarks.

If possible, it would also be nice to have a test to ensure that we don't create any System Time, except in the two place where they are currently used (Storage and CollectionHelper)

lukstbit commented:
@mikehardy Now everything should be ok. In the latest changes:

- I changed SystemTime to introduce a no arguments constructor. This is because Lint, apparently, doesn't find the instantiation of classes which rely on the default auto generated constructor
- I added a new Lint rule to find the direct instantiation of SystemTime outside of specific classes as pointed by @Arthur-Milchior 
- I changed some names according to the discussions above and to make them more consistent
- added more tests

I think this is ready to be merged. If further descriptions/explanations need changes this could be done at a later point.

Arthur-Milchior commented:
I do agree with your point of view. This is already a huge progress over current codebase, so I'm more than happy to merge it. While I still made some remarks above, those can be merged later if you still accept to work on them.
